### ISSUE TITLE
CUDA reductions should have independent input and output types

### ISSUE THREAD

**Issue by harrism** [rapidsai/cudf#446]
To enable higher-accuracy reductions, we should consider rewriting the internal reductions code to have independent input and accumulation/output datatypes. When summing integer types (e.g. Int8), the sum can be much larger than the input datatype can represent. So for higher accuracy one might want to use a 32-bit or 64-bit type for the accumulation and result. This is what Numpy does for `sum()` with integer types. In fact, it automatically uses the machine default integer size unless the user specifies the datatype.

https://docs.scipy.org/doc/numpy/reference/generated/numpy.sum.html
"dtype : dtype, optional
The type of the returned array and of the accumulator in which the elements are summed. The dtype of a is used by default unless a has an integer dtype of less precision than the default platform integer. In that case, if a is signed then the platform integer is used while if a is unsigned then an unsigned integer of the same precision as the platform integer is used."

**Comment by eyalroz** [comment]
I have actually already tweaked the current implementation of reductions to support this on a branch of mine with join-related primitives  (since I needed to count non-zero values with a cudf function, into a `gdf_size_type`). I only touched the existing code very slightly, but enough to expose different-type aggregations above. Of course, considering the fact that our reduction code has at least one other major undesirable feature (two phases), perhaps a larger refactoring is relevant.
